### ISSUE TITLE
Memory Test jumps around

### ISSUE THREAD

**Issue by KatieWoe** [phetsims/isotopes-and-atomic-mass#86]
Normal Test: 
Start: 27.0
1: 46.3
2: 48.3
3: 43.9
4: 94.1
5: 39.3
6: 105
7: 112
8: 90.5
9: 104
10: 42.9
---
Screen 1:
Start: 27.2
1: 24.1
2: 27.3
3: 27.4
4: 26.7
5: 26.4
6: 26.5
7: 28.3
8: 29.0
9: 28.1
10: 28.1
---
Screen 2:
Start: 27.3
1: 36.3
2: 101
3: 38.2
4: 81.6
5: 81.8
6: 79.7
7: 84.8
8: 83.6
9: 95.9
10: 93.8
---
For https://github.com/phetsims/QA/issues/300

**Comment by jbphet** [comment]
@KatieWoe - can you please note the version tested and also see if this is a new issue or if it exists on the published version?  When done, please assign back to me.

**Comment by KatieWoe** [comment]
Sim Tested: https://phet-dev.colorado.edu/html/isotopes-and-atomic-mass/1.1.0-rc.1/phet/isotopes-and-atomic-mass_en_phet.html
I did a quick check of the second screen on the published version, and didn't see it.  I need to wait until tomorrow to do the full tests, so I'll be sure then.

**Comment by KatieWoe** [comment]
Here are the results of the memory test for the published version.  The first is taken upon loading the sim and the rest taken every minute.
![memorytestpublished](https://user-images.githubusercontent.com/41024075/55636574-59785700-5780-11e9-9911-c40dd41ce664.png)

**Comment by jbphet** [comment]
I've spent a bit of time looking at this, and I think I have at least a cursory idea of what is going on.  The problem occurs only on the 2nd screen as [the initial bug report](https://github.com/phetsims/isotopes-and-atomic-mass/issues/86#issue-429370113) indicates.  Through testing, I've determined that the high memory values only occur when displaying "Nature's Mix" on this screen.  I can get similar behavior, but with lower memory numbers, on the currently published version.  The reason that the memory goes up so much when displaying "Nature's Mix" is that 1000 `MoveableAtom` instances are allocated, one for each of the atoms shown.  Each of these instances has several Axon `Property` instances.  When the 1.0 branch for this sim was created back in May of 2016 - almost 3 years ago - Axon `Property` instances consumed less memory.  The addition of lots of new features, many related to phet-io, have caused the per-`Property` memory usage to increase significantly, and if you multiply that increase by 1000 instances, you have a big increase in memory consumption.

I remember being a bit concerned about the approach that was being used for "Nature's Mix" when @aadish was working on this portion of the sim, but he and I measured the memory usage, and it was okay.  Now it's not, so I should probably fix it.  My initial thought is that a simpler type could be created to represent the atoms shown in the "Nature's Mix" view, and that could be used instead of `MovableAtom`.  I *think* this would significantly reduce the memory usage, and I *think* that much of the functionality of `MoveableAtom` is not needed to display "Nature's Mix".  I'll do a bit more investigation, and possibly try it out, and see how it goes.

**Comment by jbphet** [comment]
This is what the screen looks like when displaying "Nature's Mix", which is when memory consumption peaks:
![image](https://user-images.githubusercontent.com/3082831/56071004-9af09f80-5d48-11e9-8d7c-c1cdc15e0347.png)

**Comment by jbphet** [comment]
I spent a couple hours trying out something fairly simple that I thought might be a good way to resolve this problem, but it didn't work out.  I was adding too much complexity to the code and, though I potentially could have gotten it working, the code would have been much more complex and hard to maintain.

I think the best way to solve this problem is to create a base class for `MoveableAtom` called `TestChamberAtom` that has only the properties needed in the minimal view, and then have `MovableAtom` add in the other functionality needed, and then add the simple or more complex but less abundant atoms to the test chamber as appropriate.  This is not a trivial change, so I'll have to get to it another time.

**Comment by samreid** [comment]
Here's a quick change that could save around 20MB:

<details>

```
Index: js/model/NumberAtom.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- js/model/NumberAtom.js	(revision d14c5d13e0e4687e774ca9b96e44c2c4147b122b)
+++ js/model/NumberAtom.js	(date 1555611864000)
@@ -20,6 +20,8 @@
   var shred = require( 'SHRED/shred' );
   var Tandem = require( 'TANDEM/Tandem' );
 
+  const NumberDerivedProperty = DerivedPropertyIO( NumberIO );
+
   /**
    * @param {Object} options
    * @constructor
@@ -57,7 +59,7 @@
       }, {
         tandem: options.tandem.createTandem( 'chargeProperty' ),
         numberType: 'Integer',
-        phetioType: DerivedPropertyIO( NumberIO )
+        phetioType: NumberDerivedProperty
       }
     );
 
@@ -67,7 +69,7 @@
       }, {
         tandem: options.tandem.createTandem( 'massNumberProperty' ),
         numberType: 'Integer',
-        phetioType: DerivedPropertyIO( NumberIO )
+        phetioType: NumberDerivedProperty
       }
     );
 
@@ -77,7 +79,7 @@
       }, {
         tandem: options.tandem.createTandem( 'particleCountProperty' ),
         numberType: 'Integer',
-        phetioType: DerivedPropertyIO( NumberIO )
+        phetioType: NumberDerivedProperty
       }
     );
 

```
</details>

**Comment by jbphet** [comment]
I'm going to define two repeatable test procedures that demonstrate the large memory usage.  I'm going to list these in detail so that I, the QA team, and anyone else can methodically test changes.  

For the results shown below, this testing was done on Win 10, Chrome 73.0.3683.103.

Setup: There should be no other incognito windows up at the start of this test, and the test window must be closed once the test is complete.  Otherwise, I've seen different results.

### Procedure 1:
1. Open a new Chrome incognito window
2. Open the dev tools
3. Load the version of IAAM to be tested by pasting the URL in the address bar (so that no other sites are visited prior to the sim)
4. Go to the "Mixtures" screen
8. Select Sulfur (S) in the periodic table
9. Press the "Nature's Mix" radio button
10. Take a heap snapshot

#### Results
+ On v1.1.0-rc.1 (the version against which this issue was originally reported), the memory usage measured 80.8 MB, which is high for such little usage
+ On the currently published version, which is v1.0.15, the memory usage is 35.1 MB
+ On a built version of the master branch after the changes in the commits just above, the memory usage is 48.3 MB.

### Procedure 2:
1. Open a new Chrome incognito window
2. Open the dev tools
3. Load the version of IAAM to be tested by pasting the URL in the address bar (so that no other sites are visited prior to the sim)
4. Go to the "Mixtures" screen
5. Switch from adding isotopes via buckets to adding them via sliders by pressing the slider icon just below the isotope chamber
6. Move both sliders all the way to the right to add the max number of hydrogen atoms
7. Go through the first two rows of the periodic table, i.e. from H to Ne, and add the max number of atoms in this manner
8. Select Sulfur (S) in the periodic table
9. Press the "Nature's Mix" radio button
10. Take a heap snapshot

#### Results
+ On v1.1.0-rc.1 (the version against which this issue was originally reported), the memory usage measured 184.0 MB.  This does indeed seem problematic, since anything over 100 MB would crash Safari tabs on iPad 2 devices.
+ On the currently published version, which is v1.0.15, the memory usage is 69.4 MB.
On a built version of the master branch after the changes in the commits just above, the memory usage is 92.3 MB.

Since the changes I recently made keep the memory below 100 MB in these test cases, I feel that this is a sufficient improvement to this fixed and move on with a new RC.

**Comment by jbphet** [comment]
@samreid's change made sense, but I felt like we would get an even greater improvement if I were to factor our usage of mutable values in the isotopes used in the test chamber on the 2nd screen.  Based on the data logged in the previous comment, it looks like that is true, so I'm going with that approach.

**Comment by jbphet** [comment]
@samreid and @zepumph - you were assigned this issue to see if there is a PhET-iO action that should be taken to fix things.  I have a solution in place for the immediate problem and am unassigning myself, I'll leave it up to you whether to also unassign (and perhaps create a separate PhET-iO issue if you feel any other followup is warranted).

**Comment by KatieWoe** [comment]
These are the results of a quick normal memory test I did as a comparison:
Start: 26.5
1: 48.7
2: 60.4
3: 60.5
4: 61.3
5: 64.3
6: 40.8
7: 60.5
8: 62.6
9: 64.4
10: 63.9

**Comment by KatieWoe** [comment]
Procedure 1: 50.1
Procedure 2: 96.7
In addition, I did a modified version of Prcedure 2 where I did step 7 for all of the available elements.  These results were 139.

**Comment by jbphet** [comment]
@samreid and @zepumph - please close if you feel like no other PhET-iO work is necessary, or if you decide to open a seaparte issue.

**Comment by zepumph** [comment]
@samreid I'm good to close if you are.
